programmieren-lernen/dependency-injection.md

```diff
@@ -6,26 +6,39 @@ alias: ["di"]
 Dependency Injection by Example:
 ```java
 public class MeineApp {
+  // Hier übergibt man die Instanz der MeineApp Klasse
   private final DepA a = new DepA(this);
-  private final Depb b = new DepB(a);
+  // Die DepB Klasse greift auf die DepA Klasse zu
+  public final DepB b = new DepB(a);
+  private DepC c;
   public void start(String[] args) {
-    // Do Stuff
+    b.doThat();
+    // Info: Je nach umgebung, ist es manchmal nicht möglich bei Deklarationen zu übergeben,
+    // sondern erst z.b. einer onEnable Methode zum Beispiel.
+    c = new DepC(b);
   }
-  // Getters here
 }
 ```
 ```java
 public class DepA {
   private final MeineApp app;
+  // Hier kann man dann die Instanz der MeineApp Klasse verwenden, um in dem Fall um auf b zuzugreifen.
   public DepA(MeineApp app) { this.app = app; }
-  public void doThis() {}
+  // Man muss sich aber bemühen nur die Sachen zu übergeben, die auch verwendet werden.
+  // Hier also eher eine DepB Instanz, aber da diese nach DepA initialisiert wird, kann man die nicht direkt übergeben.
+  public void doThis() { app.b.doSomething(); }
 }
 ```
 ```java
 public class DepB {
   private final DepA a;
+  // Die DepB Klasse benötigt Zugriff auf die DepA Instanz.
+  // Daher übrgibt man diese im Konstruktor
   public DepB(DepA a) { this.a = a; }
   public void doThat() { a.doThis(); }
+  public void doSomething() { }
 }
 ```
-Info: es muss nicht unbedingt direkt beim deklarieren passieren. Prinzip ist es dass was man braucht im Konstruktor übergibt, wenn man die main braucht, übergibt man sie. Was ist mit cyclic dependencies? Da kann man zb. ne `init()` Methode machen.
+Das Prinzip von Dependency Injection ist, dass im Konstruktor übergeben wird, was man braucht. Wenn man die Instanz der Hauptklasse braucht, übergibt man sie.  
+Bei cyclic Dependencies (also wie `DepA` `DepB` braucht und andersrum) setzt man die Dependencies später oder übergibt man einen höheren Container wie `MeineApp`.
```

Comment:
Cyclic dependencies should be resolved by extracting the code used by both classes into a third one.
So you can instantiate the thirds one without any dependencies and pass it into A and B.